source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// Manages a block of raw memory for objects of type T. T must be
+// empty-constructible. This class carries extra member variables for tracking
+// size, and a write-pointer to support safe appends.
+//
+// MemBlockBuilder is used to safely write blocks of data into a memory
+// buffer. Due to two extra member variables, it is not optimal for storing in
+// data structures. The intended usage model is to release the raw assembled
+// memory block from the MemBlockBuilder for efficient storage.
+//
+// The goal for this class is to provide a usage model to replace direct usage
+// of memcpy with a pattern that is easy to validate for correctness by
+// inspection, asserts, and fuzzing, but when compiled for optimization is
+// roughly as efficient as raw memcpy.
+template <class T> class MemBlockBuilder {
+public:
+  // Constructs a MemBlockBuilder of the specified size.
```

Comment:
"Size" is a little ambiguous here, since it refers to the count of T objects not size in some absolute unit of memory which a name like `MemBlockBuilder` might imply absent context. You could call this "count" or just add a bit more context to the comment.